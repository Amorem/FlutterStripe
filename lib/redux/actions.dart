import 'dart:convert';
import '../models/product.dart';
import 'package:http/http.dart' as http;

import '../models/app_state.dart';
import '../models/user.dart';
import 'package:redux_thunk/redux_thunk.dart';
import 'package:redux/redux.dart';
import 'package:shared_preferences/shared_preferences.dart';

/* User Actions */
ThunkAction<AppState> getUserAction = (Store<AppState> store) async {
  final prefs = await SharedPreferences.getInstance();
  final String storedUser = prefs.getString('user');
  final user =
      storedUser != null ? User.fromJson(json.decode(storedUser)) : null;
  store.dispatch(GetUserAction(user));
};

ThunkAction<AppState> logoutUserAction = (Store<AppState> store) async {
  final prefs = await SharedPreferences.getInstance();
  await prefs.remove('user');
  User user;
  store.dispatch(LogoutUserAction(user));
};

class GetUserAction {
  final User _user;
  User get user => this._user;
  GetUserAction(this._user);
}

class LogoutUserAction {
  final User _user;
  User get user => this._user;
  LogoutUserAction(this._user);
}

/* Products Actions */
ThunkAction<AppState> getProductsAction = (Store<AppState> store) async {
  http.Response response = await http.get('http://localhost:1337/products');
  final List<dynamic> responseData = json.decode(response.body);
  List<Product> products = [];
  responseData.forEach((productData) {
    final Product product = Product.fromJson(productData);
    products.add(product);
  });
  store.dispatch(GetProductsAction(products));
};

class GetProductsAction {
  final List<Product> _products;
  List<Product> get products => this._products;
  GetProductsAction(this._products);
}

/* Cart Products Actions */

ThunkAction<AppState> toggleCartProductAction(Product cartProduct) {
  return (Store<AppState> store) async {
    final List<Product> cartProducts = store.state.cartProducts;
    final User user = store.state.user;
    final int index =
        cartProducts.indexWhere((product) => product.id == cartProduct.id);
    bool isInCart = index > -1 == true;
    List<Product> updatedCartProducts = List.from(cartProducts);
    if (isInCart) {
      updatedCartProducts.removeAt(index);
    } else {
      updatedCartProducts.add(cartProduct);
    }
    store.dispatch(ToggleCartProductAction(updatedCartProducts));
    final List<String> cartProductsIds =
        updatedCartProducts.map((product) => product.id).toList();
    await http.put('http://localhost:1337/carts/${user.cartId}',
        body: {"products": json.encode(cartProductsIds)},
        headers: {"Authorization": "Bearer ${user.jwt}"});
  };
}

class ToggleCartProductAction {
  final List<Product> _cartProducts;
  List<Product> get cartProducts => this._cartProducts;
  ToggleCartProductAction(this._cartProducts);
}

ThunkAction<AppState> getCartProductsAction = (Store<AppState> store) async {
  final prefs = await SharedPreferences.getInstance();
  final String storedUser = prefs.getString('user');
  if (storedUser == null) {
    return;
  }
  final User user = User.fromJson(json.decode(storedUser));
  http.Response response = await http.get(
      "http://localhost:1337/carts/${user.cartId}",
      headers: {'Authorization': 'Bearer ${user.jwt}'});
  final responseData = json.decode(response.body)['products'];
  List<Product> cartProducts = [];
  responseData.forEach((productData) {
    final Product product = Product.fromJson(productData);
    cartProducts.add(product);
  });
  store.dispatch(GetCartProductsAction(cartProducts));
};

class GetCartProductsAction {
  final List<Product> _cartProducts;
  List<Product> get cartProducts => this._cartProducts;
  GetCartProductsAction(this._cartProducts);
}
