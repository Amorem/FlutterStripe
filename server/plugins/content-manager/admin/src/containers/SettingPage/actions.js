import {
  ON_CLICK_EDIT_FIELD,
  ON_CLICK_EDIT_LIST_ITEM,
  ON_CLICK_EDIT_RELATION,
} from './constants';

export function onClickEditField(fieldToEdit) {
  return {
    type: ON_CLICK_EDIT_FIELD,
    fieldToEdit,
  };
}

export function onClickEditListItem(listItemToEdit) {
  return {
    type: ON_CLICK_EDIT_LIST_ITEM,
    listItemToEdit,
  };
}

export function onClickEditRelation(relationToEdit) {
  return {
    type: ON_CLICK_EDIT_RELATION,
    relationToEdit,
  };
}