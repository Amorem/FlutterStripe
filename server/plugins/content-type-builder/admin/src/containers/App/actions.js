/*
 *
 * App actions
 *
 */
import { cloneDeep, pick, set, camelCase } from 'lodash';
import { fromJS, OrderedMap } from 'immutable';
import {
  ADD_ATTRIBUTE_RELATION,
  ADD_ATTRIBUTE_TO_EXISITING_CONTENT_TYPE,
  ADD_ATTRIBUTE_TO_TEMP_CONTENT_TYPE,
  CANCEL_NEW_CONTENT_TYPE,
  CLEAR_TEMPORARY_ATTRIBUTE,
  CLEAR_TEMPORARY_ATTRIBUTE_RELATION,
  CREATE_TEMP_CONTENT_TYPE,
  DELETE_MODEL,
  DELETE_MODEL_ATTRIBUTE,
  DELETE_MODEL_SUCCEEDED,
  DELETE_TEMPORARY_MODEL,
  GET_DATA,
  GET_DATA_SUCCEEDED,
  ON_CHANGE_NEW_CONTENT_TYPE_MAIN_INFOS,
  ON_CHANGE_ATTRIBUTE,
  ON_CHANGE_RELATION,
  ON_CHANGE_RELATION_NATURE,
  ON_CHANGE_RELATION_TARGET,
  RESET_NEW_CONTENT_TYPE_MAIN_INFOS,
  RESET_EDIT_EXISTING_CONTENT_TYPE,
  RESET_EXISTING_CONTENT_TYPE_MAIN_INFOS,
  RESET_EDIT_TEMP_CONTENT_TYPE,
  RESET_PROPS,
  SAVE_EDITED_ATTRIBUTE,
  SAVE_EDITED_ATTRIBUTE_RELATION,
  SET_TEMPORARY_ATTRIBUTE,
  SET_TEMPORARY_ATTRIBUTE_RELATION,
  SUBMIT_CONTENT_TYPE,
  SUBMIT_CONTENT_TYPE_SUCCEEDED,
  SUBMIT_TEMP_CONTENT_TYPE,
  SUBMIT_TEMP_CONTENT_TYPE_SUCCEEDED,
  UPDATE_TEMP_CONTENT_TYPE,
  ON_CHANGE_EXISTING_CONTENT_TYPE_MAIN_INFOS,
} from './constants';

export function addAttributeRelation(isModelTemporary, modelName) {
  return {
    type: ADD_ATTRIBUTE_RELATION,
    isModelTemporary,
    modelName,
  };
}

export function addAttributeToExistingContentType(contentTypeName, attributeType) {
  return {
    type: ADD_ATTRIBUTE_TO_EXISITING_CONTENT_TYPE,
    attributeType,
    contentTypeName,
  };
}

export function addAttributeToTempContentType(attributeType) {
  return {
    type: ADD_ATTRIBUTE_TO_TEMP_CONTENT_TYPE,
    attributeType,
  };
}

export function cancelNewContentType() {
  return {
    type: CANCEL_NEW_CONTENT_TYPE,
  };
}

export function clearTemporaryAttribute() {
  return {
    type: CLEAR_TEMPORARY_ATTRIBUTE,
  };
}

export function clearTemporaryAttributeRelation() {
  return {
    type: CLEAR_TEMPORARY_ATTRIBUTE_RELATION,
  };
}

export function createTempContentType() {
  return {
    type: CREATE_TEMP_CONTENT_TYPE,
  };
}

export function deleteModel(modelName, context) {
  return {
    type: DELETE_MODEL,
    modelName,
    context,
  };
}

export function deleteModelAttribute(keys) {
  return {
    type: DELETE_MODEL_ATTRIBUTE,
    keys,
  };
}

export function deleteModelSucceeded(modelName) {
  return {
    type: DELETE_MODEL_SUCCEEDED,
    modelName,
  };
}

export function deleteTemporaryModel() {
  return {
    type: DELETE_TEMPORARY_MODEL,
  };
}

export function getData() {
  return {
    type: GET_DATA,
  };
}

export function getDataSucceeded({ allModels, models }, connections) {
  const initialData = allModels.reduce((acc, current) => {
    acc[current.name] = pick(current, ['name', 'collectionName', 'connection', 'description', 'mainField']);
    const attributes = OrderedMap(buildModelAttributes(current.attributes));
    set(acc, [current.name, 'attributes'], attributes);

    return acc;
  }, {});

  return {
    type: GET_DATA_SUCCEEDED,
    initialData,
    models,
    connections,
  };
}

export function onChangeExistingContentTypeMainInfos({ target }) {
  const value = target.name === 'name' ? camelCase(target.value.trim()).toLowerCase() : target.value;

  return {
    type: ON_CHANGE_EXISTING_CONTENT_TYPE_MAIN_INFOS,
    keys: target.name.split('.'),
    value,
  };
}

export function onChangeNewContentTypeMainInfos({ target }) {
  const value = target.name === 'name' ? camelCase(target.value.trim()).toLowerCase() : target.value;

  return {
    type: ON_CHANGE_NEW_CONTENT_TYPE_MAIN_INFOS,
    keys: target.name.split('.'),
    value,
  };
}

export function onChangeAttribute({ target }) {
  const value = target.name.includes('name') ? target.value.split(' ').join('') : target.value;

  return {
    type: ON_CHANGE_ATTRIBUTE,
    keys: target.name.split('.'),
    value: value,
  };
}

export function onChangeRelation({ target }) {
  const value = target.name === 'unique' ? target.value : target.value.split(' ').join('');

  return {
    type: ON_CHANGE_RELATION,
    keys: target.name.split('.'),
    value,
  };
}

export function onChangeRelationNature(nature, currentModel) {
  return {
    type: ON_CHANGE_RELATION_NATURE,
    currentModel,
    nature,
  };
}

export function onChangeRelationTarget(model, currentModel, isEditing = false) {
  return {
    type: ON_CHANGE_RELATION_TARGET,
    currentModel,
    model,
    isEditing,
  };
}

export function resetNewContentTypeMainInfos() {
  return {
    type: RESET_NEW_CONTENT_TYPE_MAIN_INFOS,
  };
}

export function resetEditExistingContentType(contentTypeName) {
  return {
    type: RESET_EDIT_EXISTING_CONTENT_TYPE,
    contentTypeName,
  };
}

export function resetExistingContentTypeMainInfos(contentTypeName) {
  return {
    type: RESET_EXISTING_CONTENT_TYPE_MAIN_INFOS,
    contentTypeName,
  };
}

export function resetEditTempContentType() {
  return {
    type: RESET_EDIT_TEMP_CONTENT_TYPE,
  };
}

export function resetProps() {
  return {
    type: RESET_PROPS,
  };
}

export function saveEditedAttribute(attributeName, isModelTemporary, modelName) {
  return {
    type: SAVE_EDITED_ATTRIBUTE,
    attributeName,
    isModelTemporary,
    modelName,
  };
}

export function saveEditedAttributeRelation(attributeName, isModelTemporary, modelName) {
  return {
    type: SAVE_EDITED_ATTRIBUTE_RELATION,
    attributeName,
    isModelTemporary,
    modelName,
  };
}

export function setTemporaryAttribute(attributeName, isModelTemporary, modelName) {
  return {
    type: SET_TEMPORARY_ATTRIBUTE,
    attributeName,
    isModelTemporary,
    modelName,
  };
}

export function setTemporaryAttributeRelation(target, isModelTemporary, source, attributeName, isEditing) {
  return {
    type: SET_TEMPORARY_ATTRIBUTE_RELATION,
    attributeName,
    isEditing,
    isModelTemporary,
    source,
    target,
  };
}

export function submitContentType(oldContentTypeName, data, context, source) {
  const attributes = formatModelAttributes(data.attributes);
  const body = Object.assign(cloneDeep(data), { attributes });

  return {
    type: SUBMIT_CONTENT_TYPE,
    oldContentTypeName,
    body,
    source,
    context,
  };
}

export function submitContentTypeSucceeded() {
  return {
    type: SUBMIT_CONTENT_TYPE_SUCCEEDED,
  };
}

export function submitTempContentType(data, context) {
  const attributes = formatModelAttributes(data.attributes);
  const body = Object.assign(cloneDeep(data), { attributes });

  return {
    type: SUBMIT_TEMP_CONTENT_TYPE,
    body,
    context,
  };
}

export function submitTempContentTypeSucceeded() {
  return {
    type: SUBMIT_TEMP_CONTENT_TYPE_SUCCEEDED,
  };
}

export function updateTempContentType() {
  return {
    type: UPDATE_TEMP_CONTENT_TYPE,
  };
}

// utils
export const buildModelAttributes = attributes => {
  const formattedAttributes = attributes.reduce((acc, current) => {
    if (current.params.type === 'enumeration') {
      acc[current.name] = Object.assign(current.params, { enum: current.params.enum.join('\n') });
    } else if (current.params.nature === 'oneWay') {
      acc[current.name] = Object.assign(current.params, { key: '-' });
    } else {
      acc[current.name] = current.params;
    }

    acc[current.name] = fromJS(acc[current.name]);

    return acc;
  }, {});

  return formattedAttributes;
};

export const formatModelAttributes = attributes =>
  Object.keys(attributes).reduce((acc, current) => {
    const attribute = Object.keys(attributes[current]).reduce(
      (acc2, curr) => {
        const value = attributes[current][curr];

        if (
          ((curr.includes('max') || curr.includes('min')) && !value) ||
          curr === 'isVirtual' ||
          (curr === 'dominant' && !value) ||
          (curr === 'columnName' && value === '') ||
          (curr === 'targetColumnName' && value === '')
        ) {
          return acc2;
        }

        if (curr === 'plugin' && !!value) {
          acc2.params.pluginValue = value;
          acc2.params.plugin = true;
        } else if (curr === 'enum') {
          acc2.params.enum = value.split('\n');
        } else if (value !== false) {
          acc2.params[curr] = value;
        } else if (curr === 'configurable') {
          acc2.params[curr] = value;
        }

        return acc2;
      },
      { name: current, params: {} },
    );

    return acc.concat(attribute);
  }, []);
