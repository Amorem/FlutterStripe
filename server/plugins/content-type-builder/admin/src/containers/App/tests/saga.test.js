/**
 * Test  sagas
 */

/* eslint-disable redux-saga/yield-effects */
import { all, fork, takeLatest, put } from 'redux-saga/effects';
import defaultSaga, { deleteModel, getData, submitCT, submitTempCT } from '../saga';

import { deleteModelSucceeded, getDataSucceeded } from '../actions';
import {
  DELETE_MODEL,
  GET_DATA,
  SUBMIT_CONTENT_TYPE,
  SUBMIT_TEMP_CONTENT_TYPE,
} from '../constants';

const response = [
  {
    models: [
      {
        icon: 'fa-cube',
        name: 'permission',
        description: '',
        fields: 6,
        source: 'users-permissions',
      },
    ],
    allModels: [
      {
        collectionName: 'users-permissions_permission',
        connection: 'default',
        description: '',
        mainField: '',
        name: 'permission',
        attributes: [
          {
            name: 'type',
            params: { type: 'string', required: true, configurable: false },
          },
          {
            name: 'controller',
            params: { type: 'string', required: true, configurable: false },
          },
        ],
      },
    ],
  },
  { connections: ['default'] },
];

describe('CTB <App /> DeleteModel saga', () => {
  let deleteModelGenerator;

  beforeEach(() => {
    deleteModelGenerator = deleteModel({
      context: { plugins: {}, updatePlugin: jest.fn() },
      modelName: 'test',
    });
    const callDescriptor = deleteModelGenerator.next({ ok: true }).value;

    expect(callDescriptor).toMatchSnapshot();
  });

  it("should not dispatch the deleteModelSucceeded action if the server didn't restart", () => {
    deleteModelGenerator.next({ ok: false }).value;

    expect(strapi.notification.success).not.toHaveBeenCalled();
  });

  it('should dispatch the deleteModelSucceeded action if it requests the data successfully', () => {
    const putDescriptor = deleteModelGenerator.next({ ok: true }).value;

    expect(putDescriptor).toEqual(
      put(deleteModelSucceeded('test', { plugins: {}, updatePlugin: jest.fn() })),
    );
    expect(strapi.notification.success).toHaveBeenCalled();
  });

  it('should call the strapi.notification if the request fails', () => {
    const response = new Error('Some Error');
    deleteModelGenerator.throw(response).value;

    expect(strapi.notification.error).toHaveBeenCalled();
  });
});

describe('CTB <App /> GetData saga', () => {
  let getDataGenerator;

  beforeEach(() => {
    getDataGenerator = getData();
    const callDescriptor = getDataGenerator.next(response).value;

    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the getDataSucceeded action if it requests the data successfully', () => {
    const putDescriptor = getDataGenerator.next(response).value;
    const [data, { connections }] = response;

    expect(putDescriptor).toEqual(put(getDataSucceeded(data, connections)));
  });

  it('should call the strapi.notification if the request fails', () => {
    const response = new Error('Some Error');
    getDataGenerator.throw(response).value;

    expect(strapi.notification.error).toHaveBeenCalled();
  });
});

describe('defaultSaga Saga', () => {
  const defaultSagaSaga = defaultSaga();

  it('should start a task to watch for GET_DATA', () => {
    const forkDescriptor = defaultSagaSaga.next().value;

    expect(forkDescriptor).toEqual(
      all([
        fork(takeLatest, GET_DATA, getData),
        fork(takeLatest, DELETE_MODEL, deleteModel),
        fork(takeLatest, SUBMIT_CONTENT_TYPE, submitCT),
        fork(takeLatest, SUBMIT_TEMP_CONTENT_TYPE, submitTempCT),
      ]),
    );
  });
});
