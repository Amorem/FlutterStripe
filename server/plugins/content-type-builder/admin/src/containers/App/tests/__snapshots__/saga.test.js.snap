// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CTB <App /> DeleteModel saga should call the strapi.notification if the request fails 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/content-type-builder/models/test",
      Object {
        "method": "DELETE",
      },
      true,
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`CTB <App /> DeleteModel saga should dispatch the deleteModelSucceeded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/content-type-builder/models/test",
      Object {
        "method": "DELETE",
      },
      true,
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`CTB <App /> DeleteModel saga should not dispatch the deleteModelSucceeded action if the server didn't restart 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/content-type-builder/models/test",
      Object {
        "method": "DELETE",
      },
      true,
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`CTB <App /> GetData saga should call the strapi.notification if the request fails 1`] = `
Object {
  "@@redux-saga/IO": true,
  "ALL": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "/content-type-builder/models",
          Object {
            "method": "GET",
          },
        ],
        "context": null,
        "fn": [Function],
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "/content-type-builder/connections",
          Object {
            "method": "GET",
          },
        ],
        "context": null,
        "fn": [Function],
      },
    },
  ],
}
`;

exports[`CTB <App /> GetData saga should dispatch the getDataSucceeded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "ALL": Array [
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "/content-type-builder/models",
          Object {
            "method": "GET",
          },
        ],
        "context": null,
        "fn": [Function],
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "CALL": Object {
        "args": Array [
          "/content-type-builder/connections",
          Object {
            "method": "GET",
          },
        ],
        "context": null,
        "fn": [Function],
      },
    },
  ],
}
`;
